File: android/app/src/main/java/com/adhiravishankar/citiesorganizer/apis/NeighborhoodAPI.java
package com.adhiravishankar.citiesorganizer.apis;

import com.adhiravishankar.citiesorganizer.models.DetailedNeighborhood;
import com.adhiravishankar.citiesorganizer.models.Neighborhood;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface NeighborhoodAPI {

    @GET("neighborhoods")
    Call<List<Neighborhood>> getNeighborhoods();

    @FormUrlEncoded
    @POST("neighborhoods")
    Call<String> postNeighborhood(@Field("name") String name,
                                  @Field("metro_id") String metroID,
                                  @Field("city_id") String cityID,
                                  @Field("link") String link,
                                  @Field("high_school_score") Integer highSchoolScore,
                                  @Field("middle_school_score") Integer middleSchoolScore,
                                  @Field("elementary_school_score") Integer elementarySchoolScore,
                                  @Field("address") String address,
                                  @Field("notes") Integer minimumValue,
                                  @Field("notes") Integer maximumValue,
                                  @Field("notes") Integer minSqft,
                                  @Field("notes") Integer maxSqft,
                                  @Field("notes") String notes,
                                  @Field("featured_image") String featuredImage);

    @GET("neighborhoods/{neighborhood}")
    Call<DetailedNeighborhood> getNeighborhood(@Path("neighborhood") String neighborhood);

    @FormUrlEncoded
    @PUT("neighborhoods/{neighborhood}")
    Call<String> updateNeighborhood(@Path("neighborhood") String neighborhood,
                                    @Field("name") String name,
                                    @Field("metro_id") String metroID,
                                    @Field("city_id") String cityID,
                                    @Field("link") String link,
                                    @Field("high_school_score") Integer highSchoolScore,
                                    @Field("middle_school_score") Integer middleSchoolScore,
                                    @Field("elementary_school_score") Integer elementarySchoolScore,
                                    @Field("address") String address,
                                    @Field("notes") Integer minimumValue,
                                    @Field("notes") Integer maximumValue,
                                    @Field("notes") Integer minSqft,
                                    @Field("notes") Integer maxSqft,
                                    @Field("notes") String notes,
                                    @Field("featured_image") String featuredImage);

    @DELETE("neighborhoods/{neighborhood}")
    Call<String> deleteNeighborhood(@Path("neighborhood") String neighborhood);
}

File: website/src/apis/NeighborhoodAPI.ts
import ky, { KyResponse } from 'ky';

import { Neighborhood } from '../interfaces/Neighborhood';
import { fetcher } from './API';


export class NeighborhoodAPI {
  baseURL: string;

  constructor() {
    this.baseURL = process.env.BASE_URL;
  }

  async neighborhoods(): Promise<SWRResponse> {
    return useSWR(this.baseURL + '/neighborhoods', fetcher);
  }

  async getNeighborhood(id: string): Promise<SWRResponse> {
    return useSWR(this.baseURL + '/neighborhoods/' + id, fetcher);
  }

  async insertNeighborhood(neighborhood: Neighborhood): Promise<KyResponse> {
    const formData = new FormData();
    formData.set('city_id', neighborhood.CityID);
    formData.set('metro_id', neighborhood.MetroID);
    formData.set('featured_image', neighborhood.FeaturedImage);
    formData.set('link', neighborhood.Link);
    formData.set('name', neighborhood.Name);
    formData.set('high_school_score', neighborhood.HighSchoolScore.toString());
    formData.set('middle_school_score', neighborhood.MiddleSchoolScore.toString());
    formData.set('elementary_school_score', neighborhood.ElementarySchoolScore.toString());
    formData.set('address', neighborhood.Address);
    formData.set('minimum_value', neighborhood.MinimumValue.toString());
    formData.set('maximum_value', neighborhood.MaximumValue.toString());
    formData.set('min_sqft', neighborhood.MinSqft.toString());
    formData.set('max_sqft', neighborhood.MaxSqft.toString());
    formData.set('notes', neighborhood.Notes);
    return ky.post(this.baseURL + '/neighborhoods', { body: formData });
  }

  async updateNeighborhood(neighborhood: Neighborhood): Promise<KyResponse> {
    const formData = new FormData();
    formData.set('city_id', neighborhood.CityID);
    formData.set('metro_id', neighborhood.MetroID);
    formData.set('featured_image', neighborhood.FeaturedImage);
    formData.set('link', neighborhood.Link);
    formData.set('name', neighborhood.Name);
    formData.set('high_school_score', neighborhood.HighSchoolScore.toString());
    formData.set('middle_school_score', neighborhood.MiddleSchoolScore.toString());
    formData.set('elementary_school_score', neighborhood.ElementarySchoolScore.toString());
    formData.set('address', neighborhood.Address);
    formData.set('minimum_value', neighborhood.MinimumValue.toString());
    formData.set('maximum_value', neighborhood.MaximumValue.toString());
    formData.set('min_sqft', neighborhood.MinSqft.toString());
    formData.set('max_sqft', neighborhood.MaxSqft.toString());
    formData.set('notes', neighborhood.Notes);
    return ky.post(this.baseURL + '/neighborhoods/' + neighborhood.ID, { body: formData });
  }

}
